File: go/Freedom_Trail.go
package main

import "math"

func findRotateSteps(rg string, key string) int {
	m := make(map[byte][]int)
	for i := 0; i < len(rg); i++ {
		m[rg[i]] = append(m[rg[i]], i)
	}
	dp := make([][]int, len(key))
	for i := 0; i < len(key); i++ {
		dp[i] = make([]int, len(rg))
		for j := 0; j < len(rg); j++ {
			dp[i][j] = -1
		}
	}
	abs := func(n int) int {
		if n >= 0 {
			return n
		}
		return -n
	}
	var mem func(k, r int) int
	mem = func(k, r int) int {
		if k == len(key) {
			return 0
		}
		if dp[k][r] != -1 {
			return dp[k][r]
		}
		if key[k] == rg[r] {
			dp[k][r] = 1 + mem(k+1, r)
		}
		dp[k][r] = math.MaxInt32
		for _, nx := range m[key[k]] {
			dp[k][r] = min(dp[k][r], 1+min(abs(r-nx), len(rg)-nx+r, len(rg)-r+nx)+mem(k+1, nx))
		}
		return dp[k][r]
	}
	return mem(0, 0)
}

File: rust/Freedom_Trail.rs
use std::{cmp, collections::HashMap, usize};

impl Solution {
    pub fn find_rotate_steps(rg: String, key: String) -> i32 {
        let mut m = HashMap::<char, Vec<i32>>::new();
        for (i, c) in rg.chars().enumerate() {
            m.entry(c).or_insert_with(Vec::new).push(i as i32);
        }
        let (mut prev, mut cur): (Vec<i32>, Vec<i32>) =
            (vec![0; rg.len() + 1], vec![0; rg.len() + 1]);
        for k in (0..key.len() - 1).rev() {
            for r in (0..rg.len() - 1).rev() {
                if k > 0 && rg.as_bytes()[r] != key.as_bytes()[k - 1] {
                    continue;
                }
                if rg.as_bytes()[r] == key.as_bytes()[k] {
                    cur[r] = 1 + prev[r];
                    continue;
                }
                cur[r] = i32::MAX;
                for nx in m.get(&key.chars().nth(k).unwrap()).unwrap().iter() {
                    cur[r] = cmp::min(
                        cur[r],
                        1 + cmp::min(
                            i32::abs(r as i32 - nx),
                            cmp::min(
                                rg.len() as i32 - r as i32 + nx,
                                rg.len() as i32 - nx + r as i32,
                            ),
                        ) + prev[*nx as usize],
                    );
                }
            }
            prev = cur.to_vec();
        }
        return cur[0];
    }
}
